pkg/policy/types/policyentry.go

```diff
@@ -0,0 +1,113 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package types
+
+import (
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+// PolicyEntry specifies the L3/L4 details of a single policy rule
+//
+// +deepequal-gen=true
+// +deepequal-gen:private-method=true
+type PolicyEntry struct {
+	// Authentication specifies the cryptographic authentication required for the traffic to be allowed
+	Authentication *api.Authentication
+
+	// Log specifies custom policy-specific Hubble logging configuration.
+	Log api.LogConfig
+
+	// EndpointSelector specifies the endpoint that this rule applies to
+	EndpointSelector api.EndpointSelector
+
+	// L3 specifies the source/destination endpoints or all endpoints if empty
+	//
+	// +deepequal-gen=false
+	L3 EndpointSelectorInterfaceSlice
+
+	// L4 specifies the source/destination port rules or none if empty
+	L4 api.PortRules
+
+	// Labels stores optional metadata.
+	Labels labels.LabelArray
+
+	// DefaultDeny is true if affected subjects should have non-selected traffic denied
+	DefaultDeny bool
+
+	// Deny is true if this rule should deny traffic
+	Deny bool
+
+	// Ingress is true if rule should affect ingress traffic, false otherwise
+	Ingress bool
+
+	// Node is true if EndpointSelector refers to a node
+	Node bool
+}
+
+func (in *PolicyEntry) DeepEqual(other *PolicyEntry) bool {
+	if other == nil {
+		return false
+	}
+	if len(in.L3) != len(other.L3) {
+		return false
+	}
+	for i := range in.L3 {
+		switch v1 := in.L3[i].(type) {
+		case api.EndpointSelector:
+			if v2, ok := other.L3[i].(api.EndpointSelector); !ok || !v1.DeepEqual(&v2) {
+				return false
+			}
+		case api.CIDR:
+			if v2, ok := other.L3[i].(api.CIDR); !ok || v1 != v2 {
+				return false
+			}
+		case api.CIDRRule:
+			if v2, ok := other.L3[i].(api.CIDRRule); !ok || !v1.DeepEqual(&v2) {
+				return false
+			}
+		default:
+			return false
+		}
+	}
+	return in.deepEqual(other)
+}
+
+// PolicyEntries is a slice of pointers to PolicyEntry
+type PolicyEntries []*PolicyEntry
+
+// EndpointSelectorInterface is a generic representation of an endpoint selector.
+// It can be converted to an EndpointSelector or FQDNSelector.
+type EndpointSelectorInterface interface {
+	IsEndpointSelectorInterface()
```

Comment:
Should this just use an actual real function that these types satisfy, such as `GetAsEndpointSelectors()`?
```suggestion
	GetAsEndpointSelectors()
```

Reply:
GetAsEndpointSelector() is a method of the slice types (i.e. EndpointSelectorSlice, FQDNSlice, CIDRSlice), while this interface is supposed to represent an individual endpoint selector. The only common method they have is String(), which seems pretty generic.

Reply:
I poked around a little in the usage of `FromPeerSelectorSlice()` and the conclusion I came to is just that it's a generic box to hide implementation details of various selectors, but different users are unpacking and doing different implementation-specific things with the resulting selectors so it's not as simple as just picking a good common function or for example adding a `GetAsEndpointSelectors()` method on all the relevant types.

I'm left wondering whether the generic could be somehow improved or taken further as this pattern "feels" a bit off, but maybe I'm biting off more than is worth chewing right now. This is something we can leave for a future iteration.